situp-plugins/elasticsearch/src/main/java/com/amazon/situp/plugins/sink/elasticsearch/BulkRetryStrategy.java

```diff
@@ -0,0 +1,141 @@
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.bulk.BackoffPolicy;
+import org.elasticsearch.action.bulk.BulkItemResponse;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.rest.RestStatus;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+public final class BulkRetryStrategy {
+    private static final Set<Integer> NON_RETRY_STATUS = new HashSet<>(
+            Arrays.asList(
+                    RestStatus.BAD_REQUEST.getStatus(),
+                    RestStatus.NOT_FOUND.getStatus(),
+                    RestStatus.CONFLICT.getStatus()
+            ));
+
+    private final RequestFunction<BulkRequest, BulkResponse> requestFunction;
+    private final BiConsumer<DocWriteRequest<?>, Throwable> logFailure;
+    private final Supplier<BulkRequest> bulkRequestSupplier;
+
+    public BulkRetryStrategy(final RequestFunction<BulkRequest, BulkResponse> requestFunction,
+                             final BiConsumer<DocWriteRequest<?>, Throwable> logFailure,
+                             final Supplier<BulkRequest> bulkRequestSupplier) {
+        this.requestFunction = requestFunction;
+        this.logFailure = logFailure;
+        this.bulkRequestSupplier = bulkRequestSupplier;
+    }
+
+    public void execute(final BulkRequest bulkRequest) throws InterruptedException {
+        // Exponential backoff run forever
+        // TODO: replace with custom backoff policy setting including maximum interval between retries
+        final BackOffUtils backOffUtils = new BackOffUtils(
+                BackoffPolicy.exponentialBackoff(TimeValue.timeValueMillis(50), Integer.MAX_VALUE).iterator());
+        handleRetry(bulkRequest, null, true, backOffUtils);
+    }
+
+    public boolean canRetry(final BulkResponse response) {
+        for (final BulkItemResponse bulkItemResponse : response) {
+            if (bulkItemResponse.isFailed() && !NON_RETRY_STATUS.contains(bulkItemResponse.status().getStatus())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean canRetry(final Exception e) {
+        return (e instanceof IOException ||
+                (e instanceof ElasticsearchException &&
+                        !NON_RETRY_STATUS.contains(((ElasticsearchException) e).status().getStatus())));
+    }
+
+    private Tuple<BulkRequest, BulkResponse> retry(final BulkRequest request) throws Exception {
+        final BulkResponse response = requestFunction.apply(request);
+        return Tuple.tuple(request, response);
+    }
+
+    private void handleRetry(
```

Comment:
1. Not sure if we need a tuple? we already have reference of the request/response of each recursive call.
2. We can avoid repeated calling of backofUtils.hasNext() by letting the next recursive call doing the check.
3. What is the issue with the recursion depth?

Check the below code?

```
        //if we move the initialAttempt as part of backOffUtils, basically bacofffUtils can return next() without any waittime the first time.
        if (backOffUtils.hasNext()) {
            backOffUtils.next();
            final BulkRequest newRequest = createBulkRequestForRetry(request, response);
            try {
                final Bulkresponse newResponse = requestFunction.apply(newRequest);
                if (newResponse.hasFailures()) {
                    //First handle non retryable with DLQ
                    handleFailures(newRequest, newResponse);
                    if (canRetry(newResponse)) {
                        handleRetry(newRequest, newResponse, backOffUtils);
                    }
                }
            } catch (final Exception e) {
                if (canRetry(e)) {
                    handleRetry(bulkRequestForRetry, false, backOffUtils);
                } else {
                    handleFailures(newRequest, e);
                }
            }
            
        } else {
            Log.error("No next attempt, this shouldn't happen");
        }
        return;
```

Reply:
> 1. Not sure if we need a tuple? we already have reference of the request/response of each recursive call.

Thanks for the catching. Yeah, it is unnecessary now.

> 2. We can avoid repeated calling of backofUtils.hasNext() by letting the next recursive call doing the check.

Makes sense. that is redundant.

> 3. What is the issue with the recursion depth?

I am not sure if in Java there is a maximum recursion depth limit that if recursion depth exceeds it will throw an Exception.

Reply:
I would not worry about memory, because on exception we actually reuse the same bulkRequest object reference between recursive calls, which is the only realistic possible never ending recursion scenario.